generator/src/main/java/com/graphicsfuzz/generator/mutateapi/ExpressionGenerator.java

```diff
@@ -0,0 +1,632 @@
+
+    globalFactManager.addFunctionFact(value, new FunctionFact(functionPrototype, argumentValues,
+        value));
+    final List<Expr> args = generateFunctionArgs(factManager, currentFunction, stmt,
+        argumentValues);
+    return new FunctionCallExpr(functionName, args);
+  }
+
+  private List<? extends Type> getAvailableTypes() {
+    // We will have to consider supporting more types but at the moment as we are using
+    // LiteralFuzzer to generate literal expressions so we are unable to cover all basic type now.
+    // The following are types currently supported by LiteralFuzzer.
+    return Arrays.asList(BasicType.BOOL, BasicType.INT, BasicType.FLOAT, BasicType.VEC2,
+        BasicType.VEC3, BasicType.VEC4);
+  }
+
+  private Value fuzzValue(Type type) {
+    // An Unknown value variable is the variable whose value could be anything. The fact manager
+    // could generate any arbitrary value but with the correct type.
+    final boolean isUnknown = generator.nextBoolean();
+    if (type == BasicType.BOOL) {
+      return new BooleanValue(
+          isUnknown ? Optional.empty() : Optional.of(generator.nextBoolean()));
+    }
+    if (BasicType.allScalarTypes().contains(type)) {
+      if (isUnknown) {
+        return new NumericValue((BasicType) type, Optional.empty());
+      }
+      if (type == BasicType.INT) {
+        final String intString = String.valueOf(generator.nextInt(INT_MAX - INT_MIN) + INT_MIN);
+        return new NumericValue(BasicType.INT, Optional.of(Integer.valueOf(intString)));
+      }
+      if (type == BasicType.FLOAT) {
+        final String floatString = randomFloatString();
+        return new NumericValue(BasicType.FLOAT, Optional.of(Float.valueOf(floatString)));
+      }
+      if (type == BasicType.UINT) {
+        final String intString = String.valueOf(generator.nextInt(INT_MAX - INT_MIN) + INT_MIN);
+        return new NumericValue(BasicType.UINT, Optional.of(Math.abs(Integer.valueOf(intString))));
+      }
+      throw new RuntimeException("Not implemented yet!");
+    }
+
+    if (isUnknown) {
+      return new CompositeValue(type, Optional.empty());
+    }
+    if (BasicType.allBasicTypes().contains(type)) {
+      final List<Value> values = new ArrayList<>();
+      for (int i = 0; i < ((BasicType) type).getNumElements(); i++) {
+        values.add(fuzzValue(((BasicType) type).getElementType()));
+      }
+      return new CompositeValue(type, Optional.of(values));
+    }
+
+    // TODO: Support struct and array type.
+    throw new RuntimeException("Not implemented yet!");
+  }
+
+  /**
+   * This function returns a pair of two numbers whose sum is equal to the given value.
+   *
+   * @param value the original value that will be split into two numbers.
+   * @return if value is unknown, returns a pair of empty value. Otherwise find and return two
+   *     numbers that will add up to the given value.
+   */
+  public Pair<Optional<Number>, Optional<Number>> getPairSum(Value value) {
```

Comment:
It would be nice to involve the fact manager here, to bias towards choosing at least one number for which we have a known fact.

E.g., if we have facts about variables or functions that compute 12, 16, 3, 6, 2, and we're trying to generate 100, then it would be quite nice if we have a reasonable change of choosing 12 + 88, or 94 + 6, or whatever.

For now feel free to file a GitHub issue for this and do TODO(issue url) here.